### ISSUE TITLE
System crash with Navi GPU and auto-detect

### ISSUE THREAD

**Issue by jeverett0902** [mayankk2308/purge-wrangler#44]
### System Configuration
Describe your system configuration here. State any anomalies (USB drives, etc.) involved as well.
- **Mac**: MacBook Pro 13" Mid-2012
- **macOS Version**: 10.15.2
- **System GPUs**: Intel HD Graphics 4000 1536 MB
- **eGPU**: Akitio Node Pro + Gigabyte Gaming AMD Radeon RX 5700 XT, 
- **Script Version**: PurgeWrangler (6.2.2)
- **Script Logs**: Attach logs generated by the script.
[pwlog-2020-01-12-17-23-13.zip](https://github.com/mayankk2308/purge-wrangler/files/4051740/pwlog-2020-01-12-17-23-13.zip)

### Issue Description
Script seems to detect eGPU, and attempts to load drivers. Soon after "Patching for AMD eGPUs..." and "Analyzing system", however, system spontaneously reboots. Reboot attempts with eGPU connected will crash and boot loop. Disconnecting eGPU will allow system to eventually restart. Plugging in eGPU will cause system crash again.
The Akitio Node Pro is not an Apple Supported eGPU configuration, but this particular eGPU solution does work fine on my Thunderbolt 3 equipted, 2018 MacBook Pro 15".
Also, the Apple TB3-TB2 adapter and Thunderbolt 2 cable I'm using to connect the Node Pro to the Mid-2012 MBP works connecting fine connecting the Node Pro to downstream TB2 devices when I use the Node Pro connected via TB3 to my 2018 MBP.

### Expected Behavior
Animated GIF on wiki suggests script should "Sanitize"/Complete installation, then prompt for reboot. eGPU should work once rebooted, or at least (per the complications table) when plugging in eGPU after boot.

### Screenshots
Here's 3 screens of what I see when running script
(Console above command prompt, if that helps):
![image](https://user-images.githubusercontent.com/6473534/72227215-d8988800-3567-11ea-8c3a-5e05d76c5458.png)
![image](https://user-images.githubusercontent.com/6473534/72227228-0087eb80-3568-11ea-98bb-a93dc512a698.png)
![image](https://user-images.githubusercontent.com/6473534/72227232-0978bd00-3568-11ea-9e01-8f414c98556b.png)
Crash and boot loop occurs after this last screen.

Here's the report from the Apple crash utility after reboot:
panic(cpu 2 caller 0xffffff802726520a): Kernel trap at 0xffffff7faac05c11, type 14=page fault, registers:
...
      Kernel Extensions in backtrace:
         com.apple.kext.AMDRadeonX6000Framebuffer(3.0.4)[9AB01170-DAF7-3848-8E2B-7428D56A860E]@0xffffff7faab5f000->0xffffff7faaf67fff
            dependency: com.apple.iokit.IOACPIFamily(1.4)[3467C105-9354-3BD2-85F3-E5AEF94BD275]@0xffffff7fa8265000
            dependency: com.apple.iokit.IOPCIFamily(2.9)[1286D5E5-A6A1-3C44-A244-04C068903DB2]@0xffffff7fa7b2f000
            dependency: com.apple.iokit.IOGraphicsFamily(569.3)[7B5CA52B-0BD1-34AB-BFC9-7D5C3A0DBE7C]@0xffffff7fa84fc000
            dependency: com.apple.AppleGraphicsDeviceControl(4.5.21)[EF326A60-9936-3267-A6E7-0044C7B4A952]@0xffffff7fa8811000
 
BSD process name corresponding to current thread: kernel_task
 
Mac OS version:
19C57
 
Kernel version:
Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64
...
System model name: MacBookPro9,2 (Mac-6F01561E16C75D06)
System shutdown begun: NO
Panic diags file available: YES (0x0)
 
System uptime in nanoseconds: 12194268926
last loaded kext at 11394918254: @kext.AMDRadeonX6000HWLibs        1.0 (addr 0xffffff7fab1da000, size 41893888)

**Comment by mayankk2308** [comment]
Please try setting up your system with the script without connecting the eGPU. Press ESC to skip script countdown. Don't enable Ti82, and select AMD when prompted. If you still see this issue, please verify that the GPU is functional. This symptom is very common if the GPU has failed/is defective. I've experienced this with two GPUs already.

**Comment by jeverett0902** [comment]
Thanks, skipping the detection worked great!

&gt;&gt; **Setup eGPU**
**Plug-in eGPU**. Press ESC if you are not plugging in eGPU.
eGPU detection skipped. Please provide more information.
**Backing up**...
Backup refreshed.
Skipping Ti82 support.
&gt;&gt; **eGPU Vendor**
 1. AMD
 2. NVIDIA
 3. Cancel
**Patching for AMD eGPUs**...
Skipping legacy kext.
Patches applied.
**Analyzing system**...
No anomalies expected.
Sanitizing system...
System sanitized.
**Modifications complete**.
**Reboot** to apply changes.

# After rebooting, and inserting eGPU:

## Disconnect icon:
![image](https://user-images.githubusercontent.com/6473534/72229984-929bee00-3580-11ea-95eb-7c79f8909b93.png)

## GPU in System Profiler:
![image](https://user-images.githubusercontent.com/6473534/72230007-b7906100-3580-11ea-9a78-6d20dd4ba21b.png)

![image](https://user-images.githubusercontent.com/6473534/72230010-c0813280-3580-11ea-977f-e523613d4965.png)

## Install latest Hashcat, and run a benchmark on a hash:

$ brew install --HEAD hashcat
$ hashcat -d 3 -b -m 1000
hashcat (v5.1.0) starting in benchmark mode...

OpenCL API (OpenCL 1.2 (Nov  9 2019 04:32:50)) - Platform #1 [Apple]
\* Device \#1: Intel(R) Core(TM) i5-3210M CPU @ 2.50GHz, skipped
\* Device \#2: HD Graphics 4000, skipped
\* Device \#3: **AMD Radeon RX 5700 XT Compute Engine, 8112/8176 MB (2044 MB allocatable), 40MCU**

Hashmode: 1000 - NTLM
Speed.\#3.........: **38563.5 MH/s** (68.08ms) @ Accel:256 Loops:1024 Thr:256 Vec:1

(38823.3 MH/s, btw, is what I got on my 2018 TB3 MacBook Pro, so the cost of the slower PCIe is trivial)

## GPU shows up for Metal support in DaVinci Resolve
![image](https://user-images.githubusercontent.com/6473534/72230313-80bb4a80-3582-11ea-8807-a8fb1719ca44.png)

## eGPU working also shows up in Activity Monitor
![image](https://user-images.githubusercontent.com/6473534/72230342-aa747180-3582-11ea-8ce4-bcff52426ef1.png)

Beautiful work! Thanks for your help! You are breathing new life into an old system!

**Comment by mayankk2308** [comment]
@jeverett0902 Great to hear all ok now. This seems to be a bug with macOS and Navi, as this does not occur for other GPU architectures (Polaris/Vega).

**Comment by alienblast** [comment]
Hi thanks for that 

I got interested because I’m having graphic problems. I don’t know if I’m going to use it.

I have an old Mac Pro 4.1 flashed to 5.1 with zero hardware upgrade.

I’m trying to fix up (not much money) trouble is it has nvidea gt120 so I’m stuck at high Sierra and I run avid protools and lots of other music production software so video graphics issues on Mojave’s and Catalina just won’t do.

But Apple don’t support metal so I’m having to learn how to upgrade the entire machine. Hence my attention to GitHub and your repro - just looking to see if anyone had edited/hacked nvidea drivers so they will work on older macs running Mojave’s or Catalina 

I’m having horrible boot problems my ego and boot rims drivers etc are really messed up so I’ve found a few solutions to fix that. But it would be amazing if you can point me in the direction of a repro that would just make a decent OS but save loads of memory and processin power so I can run lots of audio software. I run protools by avid with AAX plugins by universal audio and waves


The MacBook pro I have has intel iris pro 1536 mb but when I install Mojave’s or Catalina on that the screen flashes and the entire machine just cuts out or freezes. There’s power and fan issues and all kinds of crap.

I’m on Sierra on both machines right now researching like mad.

Anyway I’m very grateful for the email don’t worry my machines are messed up anyway and I haven’t had your script yet as I wasn’t sure if it was suitable..... bery impressive though.

Let me know if you know of drivers repels etc that will help me make both machines fast while sorting my graphics issues out on the new os - need some serious codee to fix this shit.

All the best and thanks again

**Comment by mayankk2308** [comment]
@alienblast Perhaps take a look at **dosdude**'s work: http://dosdude1.com/catalina/

**Comment by alienblast** [comment]
Yeah I’ve used dis dude but still get the graphics issues on both the nvidea GeForce gt120 on Mac Pro and the intel iris pro 1536
The MacBook with the iris is supposed to just support Catalina and metal without dissuade either way both machines are running like crap with or without dosdude.

I’m wanting to see if I can improve the performance of both machines for audio. There’s no way I can over lock my Mac pro or something like that is there? Any suggestions re graphics and speed without hardware up grade would be awesome!

I will upgrade the hardware once I have the budget but for now I embed both running. There must be a way to do this - I could do with like one command to run from GitHub like wilsonmars stuff to sort my Mac pro out as I’m a total newbie when it comes to code and I’ve totally ODd on guides. My heads burned! Lol

Thanks for your time anyway mate and btw your art work is amazing

**Comment by mayankk2308** [comment]
@alienblast ah I see what you mean now. I have no expertise in audio production. I believe good performance for audio relies heavily on the CPU. The only thing you can do is try out different audio apps (a difficult ask, of course) that more efficiently uses the CPU (Logic Pro maybe?). macOS memory management is typically very good and it tends to use as much memory as possible to operate effectively. I am out of ideas. Sorry for being unable to help!

**Comment by alienblast** [comment]
No thanks for getting back to me

As regards the nvidea graphics issue for the gt120 on my Mac Pro do you think you could code a script that will install all the bits and pieces and hacks or direct me to a guide to get it to work while running gt120 and Catalina on Mac Pro 4.1 (flashed to 5.1)

it would be awesome to get the nvidea gt120 working with the Mac Pro and as I’m no coder it would be great to get a script? I don’t know how hard it would be for you or if you can just sit there and bang one out?

I could create royalty free electronic signature music for you and your projects in return. 

I’ve been in the business over 30 years my stuff is good and I can give you some very well produced tracks for your projects and brands. I can do it anyway as you e been so helpful but I just need a time saving triple click link for terminal and brew etc or whatever to do the job. Is there a guide if you don’t have the time? Was that what all those emails you sent me were. Pretty confusing stuff is all.

See what you think I can provide you awesome value in pro production in return for you assuming I get my macs sorted out that is hahahah!!! 

Anyway no pressure man I’m up been up all night still not really getting anywhere if you have an nvidea graphics solution for Catalina on my Mac Pro with the system config isn’t eh photos I’d be so grateful. No worries if not. 

Thanks again for you kind attention and helpful friendly responses dude - you rock as does your art code

**Comment by mayankk2308** [comment]
@alienblast I'm not sure I sent you an email. You will get updates on this repository automatically if you are watching this repo (you are currently a watcher).

Anyway, the NVIDIA GT 120 will not work on Mojave or Catalina because it is not a Metal-supported device. Unfortunately there is no solution/bypass for this. I suggest sticking with Sierra/High Sierra. Try getting a used AMD Polaris GPU - should be cheap.

**Comment by alienblast** [comment]
Cheers man

**Comment by mayankk2308** [comment]
Reopening. Will attempt to address the problem if possible as I see many users affected.

**Comment by matthewhsu** [comment]
Just a note that I had the same issue as OP on a 2013 Mac Pro on 10.15.4 with Akitio Node Titan + Pulse Sapphire AMD Radeon RX 5700 XT. Installing without script detection worked for me.

**Comment by mayankk2308** [comment]
@matthewhsu v6.2.4 of the script prompts users to disconnect the eGPU after detection is complete. Does that not help?

**Comment by matthewhsu** [comment]
@mayankk2308 It did not help in my case. I tried two times: the first time, I did get the 'disconnect eGPU' message, but before I was able to disconnect it, the system crashed. I rebooted and tried the script again, but it crashed just before getting to that message.

If it's helpful I am happy to post the system log; I just need to know what to look for.

As noted previously, installing manually worked with no problem. Thank you for your hard work on this project.
